File: python/PTY_Interface.py
##
## ****  PTY_Interface Class ****
## by Aidan Macdonald (aidan.plenert.macdonald@gmail.com)
## 
## This class generates a pty device that emulates the
## tty device created by GPIB controllers such as
## Prologix GPIB-USB Controller (prologix.biz)
## 
## This class does NOT emulate the controller, but the
## controller must be attached to this PTY Interface
## 
## I may switch between the terms tty device and pty device.
## They refer to the same thing.
##


import os
import Device, Prologix

class Interface:

##%% Initialization - Interface(default_device, default_address)
## accepts arguments default_device and default_address
## the default_device will be connected at the default_address
## 
## If the Interface address (PTY_Interface.addr) is changed, the
## no longer have anything attached and will become functionless.
## UNLESS, the default device handles the address change and copies itself
## to the new address (Prologix_GPIB_USB does this, OR there is
## another device attached there.
##
## if no arguments are given at initialization, defaults are
## Device.Prologix_GPIB-USB at address 0
##
    def __init__(self,
                 default_device=Prologix.Prologix_GPIB_USB(),
                 default_address=0 ):
        self.m, self.s = os.openpty()
        self.devices = {default_address:default_device}
        self.addr = default_address

##%% printFilename()
## Prints the path to the PTY device generated (ie /dev/pts/2)
##
    def printFilename(self):
        name = os.ttyname(self.s) 
        print(name)
        return name

##%% run(readLen)
## Runs the emulator which will read all statements written from the outside
## to the pty device at the location given by self.printFilename(), then
## runs what was read through self.handleInput(tty_in) and writes output
## back to the pty device to be read from the outside
##
    def run(self, readLen=100):
        self.running = True
        while (self.running):
            tty_in = os.read(self.m, readLen)
            for tty_out in self.handleInput(tty_in):
                os.write(self.m, tty_out)


##%% handleInput(tty_in)
## Recieves raw input that was read from the pty device, and 
## breaks it up and send it to the device listed at the current
## address. That device must handle the command via dev.handleCMD(cmd, args)
##
    def handleInput(self, tty_in):
        tty_in = tty_in.strip()

        if tty_in != "" and self.addr in self.devices:
            dev = self.devices[self.addr]
            for cmd in tty_in.strip().split(";"):
                p = cmd.strip().split(" ")
                yield dev.handleCMD(self, p[0], p[1:]) + "\n"

        yield ""  # Must be a null string to prevent infinite loop

##%% addDevice(device, address)
## Preferred way of connecting a device to the interface
## at a given address
##
    def addDevice(self, dev, addr):
        self.devices[addr] = dev
        

File: c++/PTY_Interface.cpp

#include "PTY_Interface.h"
#include <pty.h>
#include <iostream>
#include <unistd.h>
#include <boost/algorithm/string.hpp>
#include <termios.h>
using namespace boost::algorithm;

static void set_noecho(int fd) { // Turn off Slave echo
  struct termios stermios;

  if (tcgetattr(fd, &stermios) < 0)
    std::cout << "Error tcgetattr()" << std::endl;

  stermios.c_lflag &= ~(ECHO | ECHOE | ECHOK | ECHONL);
  stermios.c_oflag &= ~(ONLCR);
  /* also turn off NL to CR/NL mapping on output */

  if (tcsetattr(fd, TCSANOW, &stermios) < 0)
    std::cout << "Error tcsetattr()" << std::endl;
}

Interface::Interface(Device *default_device,
		     int default_address) {
  char name[40];
  
  this->addr = default_address;
  this->addDevice(default_device, default_address);

  if (openpty(&this->m, &this->s, name, NULL, NULL) < 0)
    std::cerr << "Could not open pty device" << std::endl;
  else {
    set_noecho(this->s);
    this->pname = name;
    this->running = false;
  }
}

Interface::~Interface() {
  close(this->m);
}

string Interface::printFilename() {
  std::cout << this->pname << std::endl;
  return this->pname;
}

void Interface::run(int readLen) {
  char *tty_in = new char[readLen+1];
  vector<string> tty_out;
  string tty_full;
  int i;

  this->running = true;
  while (this->running) {
    tty_full = "";
    while (read(this->m, tty_in, readLen) >= readLen) {
      std::cout << tty_in << std::endl;
      tty_full += tty_in;
    }
    tty_full += tty_in;
    i = tty_full.find("\n");
    if ( i != string::npos )
      tty_full = tty_full.substr(0, i);

    tty_out = this->handleInput(tty_full);
    for (std::vector<string>::iterator it = tty_out.begin();
	 it != tty_out.end();
	 it++) {
      write(this->m, it->c_str(), it->length());
    }
  }

  delete[] tty_in;
}

vector<string> Interface::handleInput(string tty_in) {
  vector<string> out;
  Device *dev = NULL;
  vector<string> cmd;
  vector<string> p;
  string first;
  int i;

  trim(tty_in);
  if (tty_in != "" && this->devices.count(this->addr) > 0) {
    dev = this->devices.at(this->addr);
    if (dev == NULL) {
      out.push_back("ERROR! No Device at address: " +
		    to_string(this->addr));
    }

    boost::split(cmd, tty_in, boost::is_any_of(";\n"));
    for (i=0;i<cmd.size();i++) {
      boost::split(p, cmd.at(i), boost::is_any_of(" "));
      first = p.at(0);
      p.erase(p.begin());
      out.push_back(dev->handleCMD(this, first, &p) + "\n");
    }
  }

  out.push_back("");
  return out;
}

void Interface::addDevice(Device *dev, int addr) {
  this->devices.insert( std::pair<int, Device*>(addr, dev));
}

